### PR TITLE
RAG Tutorial Notebook

### PR DESCRIPTION
**Pull request by mihaela-bornea**
# Pull Request      

## What does this PR do?

RAG tutorial notebook and changes to simplify the RAG code.       

## Description

We include a reader factory to create generative readers with HuggingFace models, including the ones we train with PrimeQA, and OpenAI models. 
Added a RAG pipeline that includes the retriever as a searchable corpus and the Generative Reader.

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: notebooks/ir/dense/dpr_tutorial.ipynb [hunk 1]
```diff
@@ -63,12 +69,113 @@
...
+      "{\"time\":\"2023-06-19 16:37:46,980\", \"name\": \"primeqa.ir.dense.dpr_top.dpr.searcher\", \"level\": \"INFO\", \"message\": \"Using sharded faiss, reading shards from dpr_index_dir\"}\n",
+      "{\"time\":\"2023-06-19 16:37:46,982\", \"name\": \"primeqa.ir.dense.dpr_top.dpr.searcher\", \"level\": \"INFO\", \"message\": \"Reading passages_1_of_1.json.gz.records\"}\n",
+      "{\"time\":\"2023-06-19 16:37:46,993\", \"name\": \"primeqa.ir.dense.dpr_top.dpr.searcher\", \"level\": \"INFO\", \"message\": \"Using sharded faiss with 1 shards.\"}\n"
+     ]
+    }
+   ],
    "source": [
     "# Please update to point to your collection tsv format id\\ttext\\ttitle with these headers\n",
-    "path_to_collection_tsv_file=\"../path_to_tsv/sample.tsv\"\n",
+    "path_to_collection_tsv_file=\"../../../../ibm-generative-ai-cookbooks/notebooks/docs_and_qs/psgs.tsv\"\n",
```
Comment (vishwajeetkumar93) [inline notebooks/ir/dense/dpr_tutorial.ipynb:178]: Is this path hardcoded?

File: primeqa/ir/scripts/test_SC.py [hunk 2]
```diff
@@ -31,6 +31,6 @@ def select_column(data, col):
                 for rank, (ans, score) in enumerate(zip(res[q], scores[q])):
                     out.write("\t".join([queries[q]['id'], ans, str(rank+1), str(score)])+"\n")
     answers = []
-
+
```
Comment (vishwajeetkumar93) [inline primeqa/ir/scripts/test_SC.py:34]: Please remove empty lines.

File: primeqa/pipelines/qa_pipeline.py [hunk 3]
```diff
@@ -34,3 +35,29 @@ def run(self, input_texts: List[str], prefix="", suffix="", use_retriever=True):
             result[i] = i_result
 
         return result
+
```
Comment (vishwajeetkumar93) [inline primeqa/pipelines/qa_pipeline.py:38]: Please remove empty lines.

**Review by vishwajeetkumar93 — APPROVED** [review]
LGTM. I have left some comments.

File: primeqa/components/reader/reader_factory.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        contexts: List[List[str]]=None,
+        example_ids: List[str] = None,
+        *args,
+        **kwargs,
+    ) -> Dict[str, List[Dict]]:
+        return self.reader.predict(questions,
+                                    contexts=contexts,
+                                    example_ids=example_ids,
+                                    *args,
+                                    *kwargs)
```
Comment (bhavani105) [inline primeqa/components/reader/reader_factory.py:98]: Please add doctoring for the class and argument

File: primeqa/services/factories.py [hunk 5]
```diff
@@ -28,8 +28,8 @@
 
 READERS_REGISTRY = {
     ExtractiveReader.__name__: ExtractiveReader,
-    GenerativeReader.__name__: GenerativeFiDReader,
-    PromptReader.__name__: PromptFLANT5Reader,
+    GenerativeBaseReader.__name__: GenerativeFiDReader,
```
Comment (bhavani105) [inline primeqa/services/factories.py:31]: Should this be  GenerativeFidReader.__name__

**Review by bhavani105 — COMMENTED** [review]
Looks good.
Left some comments

File: docs/tutorials/02_RAG_Generator.ipynb [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "\n",
+    "pip install --upgrade pip\n",
+    "pip install primeqa"
```
Comment (bhavani105) [inline docs/tutorials/02_RAG_Generator.ipynb:38]: pip install --upgrade primeqa

**Review by avisil — APPROVED** [review]

File: tests/primeqa/services/rest_server/test_readers.py [hunk 7]
```diff
@@ -25,6 +25,6 @@ def test_get_readers(mock_client):
     assert len(readers) == 3
     assert [
         "ExtractiveReader",
-        "GenerativeReader",
-        "PromptReader",
+        "GenerativeBaseReader",
```
[the remark this answers is not in the document]
Comment (mihaela-bornea) [reply tests/primeqa/services/rest_server/test_readers.py:28]: I think I will leave the base class in there for now.

**Comment by mihaela-bornea** [comment]
I addressed all comments except:
1. Still using the GenerativeBaseReader for now since we might add another generative reader, in addition to FiD.
2. I did not update the RAG notebook since has been previously reviewed

**Review by bhavani105 — APPROVED** [review]
LGTM

**Pull request merged**